src/main/scala/com/johnsnowlabs/nlp/training/CoNLL.scala

```diff
@@ -307,14 +326,22 @@ case class CoNLL(
     val sentences = packSentence(text, doc.nerTagged)
     val tokenized = packTokenized(text, doc.nerTagged)
     val posTagged = packPosTagged(doc.posTagged)
-
     (text, docs, sentences, tokenized, posTagged, labels)
   }
 
+  private def coreTransformationWithDocId(doc: CoNLLDocument) = {
+    val docId = removeSurroundingHyphens(doc.docId.getOrElse("X"))
+    val (text, docs, sentences, tokenized, posTagged, labels) = coreTransformation(doc)
+    (docId, text, docs, sentences, tokenized, posTagged, labels)
+  }
+
   def packDocs(docs: Seq[CoNLLDocument], spark: SparkSession): Dataset[_] = {
-    import spark.implicits._
-    val rows = docs.map(coreTransformation).toDF.rdd
-    spark.createDataFrame(rows, schema)
+    val preDf = if (includeDocId) {
```

Comment:
I would keep a single version of everything, always return the column for the docId, if there's no doc ID in the data, return 0. Simplifies the code, keeps a single version of the coreTransformation() function... easier to maintain.

Reply:
That was the original idea in the very first commit. But then the optionality was required.